File: MATLAB/core_functions/compute_eigenvectors_sliding_corr.m
function [eigenvectors,eigenvalues] = compute_eigenvectors_sliding_corr(timeseries,half_window_size,n_eigen) 

% here I compute pearson correlation matrices using a sliding window. Then
% I take eigenvectors using the formula presented in the paper

% timeseries should be TxN, so every row is a timepoint and every column is
% a dimension

% here maybe we could add the case where we are full rank...?

if n_eigen > 2*half_window_size % maximum rank of corr matrix
   error('Number of requested eigenvectors is too large');
end

N = size(timeseries,2);
T = size(timeseries,1);

% matrix of eigenvectors stacked horizontally, one for each time point
eigenvectors = zeros(N,n_eigen,T-2*half_window_size); % so eigenvectors (:,:,t) are at time point t
% eigenvalues, one array of eigenvalues for each time point. I do
% T-2*half_window_size so that I exclude borders where the matrix is not
% well defined

eigenvalues = zeros(n_eigen,T-2*half_window_size);

for t=1:T-2*half_window_size

    truncated_timeseries = timeseries(t:t+2*half_window_size,:);
    zscored_truncated = zscore(truncated_timeseries); % as you can see from the formula (***) of the paper, 
                                                      % corr starts from
                                                      % summing the
                                                      % zscored
    normalizing_factor = size(truncated_timeseries,1)-1;                                              
    zscored_truncated = (1/sqrt(normalizing_factor))*zscored_truncated;
    
    % zscored_truncated contains all the vectors I am going to use for the
    % formula. So i have to scale them by 1/window_size-1. [maybe i SHALL
    % do it later at the end?]. 
    
    % ok, now that I have rearranged everything in a nice way I know that
    % the correlation matrix is zscored_truncated'*zscored_truncated.
    % However, instead of running an expensive EVD on it, now we can use
    % our formula
                                                     
    minimatrix = zscored_truncated*zscored_truncated';
    
     % Gathering eigenvectors (columns of v) and eigenvalues (diagonal of
     % a) of minimatrix. Eigenvalues will be them. Eigenvectors will be the
     % obtained coefficientsxoriginal vectors (see formula)
     
    [v,lambda] = eigs(minimatrix,n_eigen);
    
    % lambda are the eigenvalues I'm looking for:
    eigenvalues(:,t) = diag(lambda);
    
    % for the eigenvectors(see formula) I have to multiply the coefficients
    % with the original vectors
    eigenvectors(:,:,t) = zscored_truncated'*v;
   
    % normalising the eigenvectors by sqrt of eigenvalues so I can rebuild
    % matrix directly from them

    for i = 1:n_eigen
        eigenvectors(:,i,t) = eigenvectors(:,i,t)/norm(eigenvectors(:,i,t));
        eigenvectors(:,i,t) = eigenvectors(:,i,t) * sqrt(eigenvalues(i,t));
    end
end

end

File: Python/core_functions/compute_eigenvectors_sliding_corr.py
import numpy as np
from tqdm import tqdm
from scipy.linalg import eigh


def compute_eigs_corr(timeseries, n_eigen, half_window_size, ):
    """
    this function computes the pearson correlation matrices using a sliding window.
     then calculates eigenvectors using the formula presented in the paper

    :param timeseries should be TxN, so every row is a timepoint and every column is a dimension.
    :param half_window_size: half window size for sliding window
    :param n_eigen: number of eigenvectors
    :return: eigenvectors, eigenvalues
    """

    if n_eigen > 2 * half_window_size:  # maximum rank of corr matrix
        raise ValueError('Number of requested eigenvectors is too large')

    t, n = timeseries.shape
    total_iterations = t - 2 * half_window_size
    progress_bar_eigs = tqdm(total=total_iterations, desc="Calculating eigenvectors and eigenvalues:")

    eigenvectors = np.zeros((t - 2 * half_window_size, n, n_eigen))

    eigenvalues = np.zeros((n_eigen, t - 2 * half_window_size))

    for i in range(t - 2 * half_window_size):
        truncated_timeseries = timeseries[i:i + 2 * half_window_size + 1, :]
        zscored_truncated = (truncated_timeseries - np.mean(truncated_timeseries, axis=0)) / np.std(
            truncated_timeseries, axis=0)

        normalizing_factor = truncated_timeseries.shape[0] - 1
        zscored_truncated /= np.sqrt(normalizing_factor)

        # correlation matrix
        minimatrix = zscored_truncated @ zscored_truncated.T
        ns = len(minimatrix)

        # Gathering eigenvectors (columns of v) and eigenvalues (diagonal of a) of minimatrix.
        # Eigenvalues will be them. Eigenvectors will be the obtained coefficients x original vectors.
        eigenvalues_t, eigenvectors_t = eigh(minimatrix, subset_by_index=[ns - n_eigen, ns - 1],
                                             overwrite_a=True,
                                             check_finite=False)
        eigenvectors_t = np.flip(eigenvectors_t, axis=1)
        eigenvalues_t = np.flip(eigenvalues_t, axis=0)
        eigenvalues[:, i] = eigenvalues_t
        eigenvectors[i, :, :] = np.dot(zscored_truncated.T, eigenvectors_t)

        # normalizing the eigenvectors by sqrt of eigenvalues
        for j in range(n_eigen):
            eigenvectors[i, :, j] = eigenvectors[i, :, j] / np.linalg.norm(eigenvectors[i, :, j])
            eigenvectors[i, :, j] = eigenvectors[i, :, j] * np.sqrt(eigenvalues[j, i])

        progress_bar_eigs.update(1)

    return eigenvectors, eigenvalues